### PR TITLE
Make hires fix not crash with combinatorial processing

### PR DESCRIPTION
**Pull request by akx**
I'm not sure at all this is the right fix, as my eyes kinda glaze over trying to figure out what the difference between hr_prompt, hr_prompts, all_hr_prompts etc. in the webui is.

However, with this, hires fix doesn't _crash_. 

Refs https://github.com/adieyal/sd-dynamic-prompts/issues/474 (won't say it fixes it at this point).

The webui commit that caused this regression is https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/ff0e17174f8d93a71fdd5a4a80a4629bbf97f822

### CODE REVIEW

**Comment by joaopedrosgs** [comment]
It doesn't crash, but it seems that the hires parameters are wrong, because the generated image is a mess

**Comment by Panchvzluck** [comment]
For what I've tried to fix it on the SD side (https://github.com/Panchvzluck/stable-diffusion-webui/commit/08cd7a6cf4ab37e9e78368e690e6e64c9fd564f2)
It seems that
- prompt, negative prompt: the literal positive or negative prompt on the firstpass. (be a list, etc)
- all_prompts, all_negative_prompts = the literal value above but applied to the style (depending if it's none or not)
```
if type(self.prompt) == list:
            self.all_prompts = self.prompt
        else:
            self.all_prompts = self.batch_size * self.n_iter * [self.prompt]

        if type(self.negative_prompt) == list:
            self.all_negative_prompts = self.negative_prompt
        else:
            self.all_negative_prompts = self.batch_size * self.n_iter * [self.negative_prompt]

        self.all_prompts = [shared.prompt_styles.apply_styles_to_prompt(x, self.styles) for x in self.all_prompts]
        self.all_negative_prompts = [shared.prompt_styles.apply_negative_styles_to_prompt(x, self.styles) for x in self.all_negative_prompts]
```
-prompts, negative_prompts: the prompts/negative prompts applied to the batch size
```
for n in range(p.n_iter):
            p.iteration = n
            ....
            p.prompts = p.all_prompts[n * p.batch_size:(n + 1) * p.batch_size]
            p.negative_prompts = p.all_negative_prompts[n * p.batch_size:(n + 1) * p.batch_size]
```

-hr_prompt, hr_negative_prompt: the literal positive or negative hr prompt if you define it on the hr box. If not, it is the same as the firstpass.

```
        if self.hr_prompt == '':
                    self.hr_prompt = self.prompt

        if self.hr_negative_prompt == '':
            self.hr_negative_prompt = self.negative_prompt
```
-all_hr_prompts, all_hr_negative_promots: similar to the first pass. (applied the above to the style)
```
 if type(self.hr_prompt) == list:
            self.all_hr_prompts = self.hr_prompt
        else:
            self.all_hr_prompts = self.batch_size * self.n_iter * [self.hr_prompt]

        if type(self.hr_negative_prompt) == list:
            self.all_hr_negative_prompts = self.hr_negative_prompt
        else:
            self.all_hr_negative_prompts = self.batch_size * self.n_iter * [self.hr_negative_prompt]

        self.all_hr_prompts = [shared.prompt_styles.apply_styles_to_prompt(x, self.styles) for x in self.all_hr_prompts]
        self.all_hr_negative_prompts = [shared.prompt_styles.apply_negative_styles_to_prompt(x, self.styles) for x in self.all_hr_negative_prompts]
```
-all_hr_prompts, all_hr_negative_prompts: applied the above to the batch size, but this code is inside another functions (extra networks)
```
def parse_extra_network_prompts(self):
        res = super().parse_extra_network_prompts()

        if self.enable_hr:
            self.hr_prompts = self.all_hr_prompts[self.iteration * self.batch_size:(self.iteration + 1) * self.batch_size]
            self.hr_negative_prompts = self.all_hr_negative_prompts[self.iteration * self.batch_size:(self.iteration + 1) * self.batch_size]

            self.hr_prompts, self.hr_extra_network_data = extra_networks.parse_prompts(self.hr_prompts)

        return res
```

BTW, the last piece of code is what makes the prompt change in the highres fix pass (discovered here: https://github.com/adieyal/sd-dynamic-prompts/issues/435#issuecomment-1564775154). Removing that makes it work as it should, but hr prompt (pos/neg) stop working

**Comment by joaopedrosgs** [comment]
> For what I've tried to fix it on the SD side ([Panchvzluck/stable-diffusion-webui@08cd7a6](https://github.com/Panchvzluck/stable-diffusion-webui/commit/08cd7a6cf4ab37e9e78368e690e6e64c9fd564f2)) It seems that
> 
> * prompt, negative prompt: the literal positive or negative prompt on the firstpass. (be a list, etc)
> * all_prompts, all_negative_prompts = the literal value above but applied to the style (depending if it's none or not)
> 
> ```
> if type(self.prompt) == list:
>             self.all_prompts = self.prompt
>         else:
>             self.all_prompts = self.batch_size * self.n_iter * [self.prompt]
> 
>         if type(self.negative_prompt) == list:
>             self.all_negative_prompts = self.negative_prompt
>         else:
>             self.all_negative_prompts = self.batch_size * self.n_iter * [self.negative_prompt]
> 
>         self.all_prompts = [shared.prompt_styles.apply_styles_to_prompt(x, self.styles) for x in self.all_prompts]
>         self.all_negative_prompts = [shared.prompt_styles.apply_negative_styles_to_prompt(x, self.styles) for x in self.all_negative_prompts]
> ```
> 
> * prompts, negative_prompts: the all_prompts/all_negative prompts applied to the batch size
> 
> ```
> for n in range(p.n_iter):
>             p.iteration = n
>             ....
>             p.prompts = p.all_prompts[n * p.batch_size:(n + 1) * p.batch_size]
>             p.negative_prompts = p.all_negative_prompts[n * p.batch_size:(n + 1) * p.batch_size]
> ```
> 
> * hr_prompt, hr_negative_prompt: the literal positive or negative hr prompt if you define it on the hr box. If not, it is the same as the firstpass.
> 
> ```
>         if self.hr_prompt == '':
>                     self.hr_prompt = self.prompt
> 
>         if self.hr_negative_prompt == '':
>             self.hr_negative_prompt = self.negative_prompt
> ```
> 
> * all_hr_prompts, all_hr_negative_prompts: similar to the first pass. (applied the above to the style)
> 
> ```
>  if type(self.hr_prompt) == list:
>             self.all_hr_prompts = self.hr_prompt
>         else:
>             self.all_hr_prompts = self.batch_size * self.n_iter * [self.hr_prompt]
> 
>         if type(self.hr_negative_prompt) == list:
>             self.all_hr_negative_prompts = self.hr_negative_prompt
>         else:
>             self.all_hr_negative_prompts = self.batch_size * self.n_iter * [self.hr_negative_prompt]
> 
>         self.all_hr_prompts = [shared.prompt_styles.apply_styles_to_prompt(x, self.styles) for x in self.all_hr_prompts]
>         self.all_hr_negative_prompts = [shared.prompt_styles.apply_negative_styles_to_prompt(x, self.styles) for x in self.all_hr_negative_prompts]
> ```
> 
> * hr_prompts, hr_negative_prompts: applied the above to the batch size, but this code is inside another functions (extra networks)
> 
> ```
> def parse_extra_network_prompts(self):
>         res = super().parse_extra_network_prompts()
> 
>         if self.enable_hr:
>             self.hr_prompts = self.all_hr_prompts[self.iteration * self.batch_size:(self.iteration + 1) * self.batch_size]
>             self.hr_negative_prompts = self.all_hr_negative_prompts[self.iteration * self.batch_size:(self.iteration + 1) * self.batch_size]
> 
>             self.hr_prompts, self.hr_extra_network_data = extra_networks.parse_prompts(self.hr_prompts)
> 
>         return res
> ```
> 
> Just as an explanation that I could get from these variables. Not related to the crash itself.
> 
> Also, the last piece of code is what makes the prompt change in the highres fix pass (discovered here: [#435 (comment)](https://github.com/adieyal/sd-dynamic-prompts/issues/435#issuecomment-1564775154)). Removing that makes dynamic prompts/wildcards work as it should on the hiresfix pass, but hr prompt (pos/neg) stop working

I tried to remove the hr_prompts and hr_negative_prompts  lines but it started to crash again

**Comment by joaopedrosgs** [comment]
> > > ```
> > > def parse_extra_network_prompts(self):
> > >         res = super().parse_extra_network_prompts()
> > > 
> > >         if self.enable_hr:
> > >             self.hr_prompts = self.all_hr_prompts[self.iteration * self.batch_size:(self.iteration + 1) * self.batch_size]
> > >             self.hr_negative_prompts = self.all_hr_negative_prompts[self.iteration * self.batch_size:(self.iteration + 1) * self.batch_size]
> > > 
> > >             self.hr_prompts, self.hr_extra_network_data = extra_networks.parse_prompts(self.hr_prompts)
> > > 
> > >         return res
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Just as an explanation that I could get from these variables. Not related to the crash itself.
> > > Also, the last piece of code is what makes the prompt change in the highres fix pass (discovered here: [#435 (comment)](https://github.com/adieyal/sd-dynamic-prompts/issues/435#issuecomment-1564775154)). Removing that makes dynamic prompts/wildcards work as it should on the hiresfix pass, but hr prompt (pos/neg) stop working
> > 
> > 
> > I tried to remove the hr_prompts and hr_negative_prompts lines but it started to crash again
> 
> Can you try the processing.py I'm using on my fork? (https://github.com/Panchvzluck/stable-diffusion-webui/blob/dev/modules/processing.py). I basically moved the functions outside of the "StableDiffusionProcessing" class and "StableDiffusionProcessingTxt2Img" to the "Processed" function (So changed the self.variables to p.variables). Also the hr_prompts is defined on line 692 instead. Just tried with the latest commit of dynamic prompts and it works. Remember though, this processing.py makes hr positive/negative prompt to not work. The rest works as it should (dynamic prompts/wildcards and hr sampler)

It works!

**Comment by adieyal** [comment]
@akx looks reasonable - the auto1111 code is a little hard to follow.

**Comment by akx** [comment]
@adieyal Let's not merge this yet as the comments are saying it's not working. There's also an alternative patch in there.

**Comment by brimston3** [comment]
The simple fix that copies all_prompts into hr_all_prompts and again for negative worked for me with auto1111 version v1.3.0 (git 20ae71f).

The TL;DR for p.hr_prompt is it gets populated from p.prompt if it is unset, which means it gets the initial template and not the post-RNGd one. Since dynamic-prompts (correctly) only updates the p.all_prompts variable and not p.prompt, we get all the possible tokens in the hires prompts. Incidentally, the hr parameters infotext generator used by auto1111 v1.3.0 completely ignores that p.hr_all_prompts might be different per image.

I looked at the linked fork and it is moving code up and down the processing.py module into the shared img2img code where hr upscale doesn't apply. I'm not going to try that as it's unlikely to be accepted into main, which means it'll be a fragile fix.

**Comment by brimston3** [comment]
Incidentally, if you want to verify that the hr_prompt is being properly set per image, I modified auto1111's modules/processing.py and added to the end of create_infotext():

```
    hr_prompt_text = ""
    if enable_hr:
        if all_prompts[index] != p.all_hr_prompts[index]:
            hr_prompt_text += f"\nHires prompt: {p.all_hr_prompts[index]}" if p.all_hr_prompts[index] else ""
        if p.all_negative_prompts[index] != p.all_hr_negative_prompts[index]:
            hr_prompt_text += f"\nHires negative prompt: {p.all_hr_negative_prompts[index]}" if p.all_hr_negative_prompts[index] else ""

    return f"{all_prompts[index]}{negative_prompt_text}{hr_prompt_text}\n{generation_params_text}".strip()
```

**Comment by akx** [comment]
Okay, actually, let's see if this fixes things for now and try a different patch if it doesn't.
